lms/djangoapps/course_home_api/outline/views.py

```diff
@@ -375,6 +381,190 @@ def finalize_response(self, request, response, *args, **kwargs):
+    )
+    COURSE_BLOCKS_CACHE_TIMEOUT = 60 * 60  # 1 hour
+
+    def get(self, request, *args, **kwargs):
+        """
+        Get the visible course blocks (from course to vertical types) for the given course.
+        """
+        course_key_string = kwargs.get('course_key_string')
+        course_key = CourseKey.from_string(course_key_string)
+        course = get_course_or_403(request.user, 'load', course_key, check_if_enrolled=False)
+        staff_access = has_access(request.user, 'staff', course_key)
+
+        masquerade_object, request.user = setup_masquerade(
+            request,
+            course_key,
+            staff_access=staff_access,
+            reset_masquerade_data=True,
+        )
+
+        user_is_masquerading = is_masquerading(request.user, course_key, course_masquerade=masquerade_object)
+
+        allow_anonymous = COURSE_ENABLE_UNENROLLED_ACCESS_FLAG.is_enabled(course_key)
+        allow_public = allow_anonymous and course.course_visibility == COURSE_VISIBILITY_PUBLIC
+        allow_public_outline = allow_anonymous and course.course_visibility == COURSE_VISIBILITY_PUBLIC_OUTLINE
+        enrollment = CourseEnrollment.get_enrollment(request.user, course_key)
+
+        try:
+            user_cohort = get_cohort(request.user, course_key, use_cached=True)
+        except ValueError:
+            user_cohort = None
+
+        cache_key = self.COURSE_BLOCKS_CACHE_KEY_TEMPLATE.format(
+            course_key_string=course_key_string,
+            course_version=str(course.course_version),
+            user_id=request.user.id,
+            enrollment_mode=getattr(enrollment, 'mode', ''),
+            user_cohort_id=getattr(user_cohort, 'id', ''),
+            allow_public=allow_public,
+            allow_public_outline=allow_public_outline,
+            is_masquerading=user_is_masquerading,
+        )
+        if navigation_sidebar_caching_is_disabled := courseware_disable_navigation_sidebar_blocks_caching():
+            cached = False
+            course_blocks = None
+        else:
+            course_blocks = cache.get(cache_key)
+            cached = course_blocks is not None
+
+        if not course_blocks:
+            if getattr(enrollment, 'is_active', False) or bool(staff_access):
+                course_blocks = get_course_outline_block_tree(request, course_key_string, request.user)
+            elif allow_public_outline or allow_public or user_is_masquerading:
+                course_blocks = get_course_outline_block_tree(request, course_key_string, None)
+
+            if navigation_sidebar_caching_is_disabled:
+                cache.set(cache_key, course_blocks, self.COURSE_BLOCKS_CACHE_TIMEOUT)
+
+        course_blocks = self.filter_inaccessible_blocks(course_blocks, course_key)
+
+        if cached:
+            # If the data was cached, we need to mark the blocks as complete or not complete.
+            course_blocks = self.mark_complete_recursive(course_blocks)
```

Comment:
Please explain in a comment why we need to mark the blocks as complete only when we get a cache hit, and not when there's a cache miss.

Reply:
Comment updated.